crate/server/src/core/operations/rekey_keypair.rs

```diff
@@ -16,37 +21,70 @@ use crate::{
 pub async fn rekey_keypair(
     kms: &KMS,
     request: ReKeyKeyPair,
-    owner: &str,
+    user: &str,
     params: Option<&ExtraDatabaseParams>,
 ) -> KResult<ReKeyKeyPairResponse> {
     trace!("Internal rekey key pair");
 
-    let private_key_unique_identifier =
-        request
-            .private_key_unique_identifier
-            .as_ref()
-            .ok_or_else(|| {
-                KmsError::NotSupported(
-                    "Rekey keypair: ID place holder is not yet supported an a key ID must be \
-                     supplied"
-                        .to_string(),
-                )
-            })?;
-
     let attributes = request.private_key_attributes.as_ref().ok_or_else(|| {
         KmsError::InvalidRequest(
             "Rekey keypair: the private key attributes must be supplied".to_owned(),
         )
     })?;
 
+    // there must be an identifier
+    let uid_or_tags = request
+        .private_key_unique_identifier
+        .clone()
+        .ok_or(KmsError::UnsupportedPlaceholder)?;
+
+    // retrieve from tags or use passed identifier
+    let mut owm_s = kms
+        .db
+        .retrieve(&uid_or_tags, user, ObjectOperationType::Rekey, params)
+        .await?
+        .into_iter()
+        .filter(|owm| {
+            // only active objects
+            if owm.state != StateEnumeration::Active {
+                return false
+            }
+            // only private keys
+            if owm.object.object_type() != ObjectType::PrivateKey {
+                return false
+            }
+            // if a Covercrypt key, it must be a master secret key
+            if let Ok(attributes) = owm.object.attributes() {
+                if attributes.key_format_type == Some(KeyFormatType::CoverCryptSecretKey) {
+                    // a master key should have policies in the attributes
+                    return policy_from_attributes(attributes).is_ok()
+                }
+            }
+            true
```

Comment:
Maybe try writing these conditions as a boolean equation instead.

Reply:
IMHO, early returns are easier to read